### ISSUE TITLE
Revert #1535

### ISSUE THREAD

**Issue by Finomnis** [STEllAR-GROUP/hpx#1620]
I think we should revert #1535, as it can be proven that the necessity of querying a second id_type to a promise is the result of incorrect code.

I posted the detailed thought process as a comment of #1535.

**Comment by hkaiser** [comment]
@Finomnis In #1535 you said:

> If someone needs to request a second id_type directly from a promise, this means that the first id_type he requested is out-of-reach for the person requesting it, and it can not be guaranteed that it did not get released yet.

I wouldn't agree to this. I'd rather find a way to make sure the object the users calls get_id() for is still in valid state.

**Comment by sithhell** [comment]
Is this still an issue? I would also like to think that what @Finomnis describes in #1535 should always work, even the parts tagged as 'UNSAFE' as the promise should really hold the component alive. If that really is an issue, is there a testcase demonstrating this?

**Comment by Finomnis** [comment]
Testcase:

``` C++
// Copyright (c)       2015 Martin Stumpf
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <hpx/config.hpp>
#include <hpx/hpx.hpp>
#include <hpx/hpx_start.hpp>
#include <hpx/include/iostreams.hpp>

int hpx_main(int argc, char* argv[])
{
    {
        hpx::id_type promise_id;

        {

            hpx::promise<int> p;

            {
                auto local_promise_id = p.get_id();
                hpx::cout << local_promise_id << hpx::endl;
            }

            hpx::this_thread::sleep_for(boost::chrono::milliseconds(100));

            promise_id = p.get_id();
            hpx::cout << promise_id << hpx::endl;

        }

        hpx::this_thread::sleep_for(boost::chrono::milliseconds(100));

        // This segfaults, because the promise is not alive any more.
        // It SHOULD get kept alive by AGAS though.
        hpx::set_lco_value(promise_id, 10, false); 

    }
	return hpx::finalize();
}

int main(int argc, char* argv[])
{
	// initialize HPX, run hpx_main.
	hpx::start(argc, argv);

	// wait for hpx::finalize being called.
	return hpx::stop();
}
```

**Comment by hkaiser** [comment]
I still think we should rather make this test case work than to inhibit this kind of thing altogether.

**Comment by sithhell** [comment]
The problem here is in that line:
https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/lcos/promise.hpp#L189

upon the first retrieval of the GID, the credit count of the GID in the promise itself is set to zero. When the promise goes out of scope, AGAS thinks the credit is zero and issues the deletion. The promise GID needs to get split for this case in order to keep the promise itself alive.

**Comment by Finomnis** [comment]
That's not true, the gid in the promise becomes an unmanaged gid, which is
important to prevent it from keeping itself alive, causing a recursive loop
preventing any deallocation.

If the promise goes out of scope out will get kept alive by the other gid.
The problem is  if the other gid gets out of scope, then agas forgets about
the promise, even if the user issues a second gid. (base problem ist : agas
can't be "reminded" to re-learn about a gid once it is forgotten).
But as I already said, it can be proven that requesting a second gid is
never necessary, so that functionality should be removed.

**Comment by sithhell** [comment]
On 10/29/2015 12:08 PM, Martin Stumpf wrote:
> That's not true, the gid in the promise becomes an unmanaged gid, which is
> important to prevent it from keeping itself alive, causing a recursive loop
> preventing any deallocation.

`promise<T>::get_id()` returns a *managed* `id_type` which keeps the
promise alive as long as the id_type doesn't get out of scope.
The first time `promise<T>::get_id()` gets called, the returned
`id_type` takes away all the credit. Once this id_type goes out of
scope, the total global credit in agas is then set to zero.
Once this happened and the promise itself goes out of scope, it doesn't
have any remaining credits and will eventually delete itself. The second
request for an id is never accounted for in the promise itself (it's
gid's credit remains 0, aka has no credits) and it will get deleted
which then eventually leads to the segfault.


> 
> If the promise goes out of scope out will get kept alive by the other gid.
> The problem is if the other gid gets out of scope, then agas forgets about
> the promise, even if the user issues a second gid. (base problem ist : agas
> can't be "reminded" to re-learn about a gid once it is forgotten).

Right. that is why we need to *split* the credit instead of taking all
the credit with us. That way, we can request as many id_types as we want.

> But as I already said, it can be proven that requesting a second gid is
> never necessary, so that functionality should be removed.

Why is it never necessary? Can you show the prove? What should happen
instead when `promise<T>::get_id()` is being called? We can't invalidate
the promise in that case.

**Comment by Finomnis** [comment]
1. Exactly.

2. If you split the credit, there is no way to delete the promise, as it
keeps itself alive.

3. The second call to get_id should throw.

The basic idea of the proof is, that in order to prevent 1., the user would
actively have to keep one promise alive, which he could simple re-use
instead of using a new one.

**Comment by sithhell** [comment]
Fixed by merging #1828
